test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs

```diff
@@ -0,0 +1,214 @@
+            {
+                (NetworkControllerStatus networkControllerStatus, bool isWithinTolerancePeriod) =
+                    this.NetworkStatusTimeline.Expect<InvalidOperationException>(
+                        () => throw new InvalidOperationException("Results are present, but NetworkStatusTimeline is empty."))
+                    .GetNetworkControllerStatusAndWithinToleranceAt(this.ExpectedTestResults.Current.CreatedAt);
+
+                (ulong addNetOffSuccess, ulong addNetOnTolSuccess, ulong addNetOnFailure, ulong addMismatchSuccess) =
+                        this.CheckUnmatchedResult(this.ExpectedTestResults.Current, networkControllerStatus, isWithinTolerancePeriod, networkOffSuccess, networkOnToleratedSuccess, networkOnFailure, mismatchSuccess);
+                networkOffSuccess += addNetOffSuccess;
+                networkOnToleratedSuccess += addNetOnTolSuccess;
+                networkOnFailure += addNetOnFailure;
+                mismatchSuccess += addMismatchSuccess;
+                hasExpectedResult = await this.ExpectedTestResults.MoveNextAsync();
+            }
+
+            while (hasActualResult)
+            {
+                // Log message for unexpected case.
+                Logger.LogError($"[{nameof(DirectMethodReportGenerator)}] Actual test result source has unexpected results.");
+
+                mismatchFailure++;
+
+                // Log actual queue items
+                Logger.LogError($"Unexpected actual test result: {this.ActualTestResults.Current.Source}, {this.ActualTestResults.Current.Type}, {this.ActualTestResults.Current.Result} at {this.ActualTestResults.Current.CreatedAt}");
+
+                hasActualResult = await this.ActualTestResults.MoveNextAsync();
+            }
+
+            return new DirectMethodReport(
+                this.trackingId,
+                this.ExpectedSource,
+                this.ActualSource,
+                this.ResultType,
+                networkOnSuccess,
+                networkOffSuccess,
+                networkOnToleratedSuccess,
+                networkOffToleratedSuccess,
+                networkOnFailure,
+                networkOffFailure,
+                mismatchSuccess,
+                mismatchFailure);
+        }
+
+        (ulong networkOffSuccess, ulong networkOnToleratedSuccess, ulong networkOnFailure, ulong mismatchSuccess) CheckUnmatchedResult(
+            TestOperationResult testOperationResult,
+            NetworkControllerStatus networkControllerStatus,
+            bool isWithinTolerancePeriod,
+            ulong networkOffSuccess,
+            ulong networkOnToleratedSuccess,
+            ulong networkOnFailure,
+            ulong mismatchSuccess)
+        {
+            DirectMethodTestResult dmTestResult = JsonConvert.DeserializeObject<DirectMethodTestResult>(testOperationResult.Result);
+            HttpStatusCode statusCode = JsonConvert.DeserializeObject<HttpStatusCode>(dmTestResult.Result);
```

Comment:
convert it to int and use "YourEnum foo = (YourEnum)yourInt;"

Reply:
I'm going to keep it like this for now. Currently, the test result won't be deserialized properly, because it's not being serialized from the DMSender, but I think it should be serialized from the DMSender. (currently it's just a string with semicolon delimiters, but that is less desirable than a serialized json)

The DM connectivity test won't work until that is in place, but I think we should keep the Generator and its tests as is and fix the DMSender